### ISSUE TITLE
NetworkSecurityGroup (NSG) removes and recreate security rules

### ISSUE THREAD

**Issue by ie-zero** [pulumi/pulumi-azure-native#1087]
When the security rules are are added as separate resources after the creation of the `NetworkSecurityGroup` the rules are deleted and re-create when the `--refresh` option is passed. 

## Steps to reproduce
Stack: NetworkSecurityGroup with one security rule which is added as a separate `Pulumi.AzureNative.Network.SecurityRule` resource.  

1. Execute  `pulumi up --refresh`. The stack will be created as expected.

2. Execute  `pulumi up --refresh`. The security rules will be deleted.
    ``` console
	Updating (Pulumi.Bug2247):
		 Type                                          Name                               Status      Info
		 pulumi:pulumi:Stack                           Pulumi-Bug2247-Pulumi.Bug2247
		 +- azure-native:resources:ResourceGroup       ResourceGroup
		 +- azure-native:network:SecurityRule          NSG-AllowVirtualNetworkRdpInbound
	 ~   +- azure-native:network:NetworkSecurityGroup  NSG                                updated     [diff: -securityRules]
		 +- azure-native:network:NetworkSecurityGroup  NSG2
    ```
	
3. Execute  `pulumi up --refresh`. The stack will be created again. 
    ``` console
	Updating (Pulumi.Bug2247):
		 Type                                          Name                               Status      Info
		 +- azure-native:resources:ResourceGroup       ResourceGroup
		 +- azure-native:network:SecurityRule          NSG-AllowVirtualNetworkRdpInbound
	 +   +- azure-native:network:SecurityRule          NSG-AllowVirtualNetworkRdpInbound  create
		 +- azure-native:network:NetworkSecurityGroup  NSG2
    ```
.
.
.

**Expected:** No rules to be removed, if there are no changes to the state of the expected state of the resource.  
**Actual:** The security rules removed, added, removed ... on each consecutive call of the `pulumi up --refresh`

## Software Versions
Pulumi CLI = v3.10.1
.NET Core = 3.1

NuGet Packages 
Pulumi = 3.*
Pulumi.AzureNative = 1.22.0

## Example Code 
NSG presents the issue. NSG2 is actually stable and works as expected.

``` csharp
using Pulumi;
using Pulumi.AzureNative.Network;
using AzureNative = Pulumi.AzureNative;
using NetworkSecurityGroupArgs = Pulumi.AzureNative.Network.NetworkSecurityGroupArgs;

namespace Bug2247
{
    public sealed class MyStack : Stack
    {
        public MyStack()
        {
            var resourceGroup = new AzureNative.Resources.ResourceGroup("ResourceGroup", new AzureNative.Resources.ResourceGroupArgs
            {
                Location = "WestEurope",
                ResourceGroupName = "PULUMI-BUG2247".ToUpper()
            });

            var nsg = new NetworkSecurityGroup("NSG", new NetworkSecurityGroupArgs
            {
                NetworkSecurityGroupName = "NSG",
                ResourceGroupName = resourceGroup.Name,
                Location = resourceGroup.Location
            });

            _ = new SecurityRule($"NSG-AllowVirtualNetworkRdpInbound", new SecurityRuleArgs
            {
                SecurityRuleName = "AllowVirtualNetworkRdpInbound",
                Priority = 120,
                Direction = SecurityRuleDirection.Inbound,
                Access = SecurityRuleAccess.Allow,
                Protocol = SecurityRuleProtocol.Tcp,
                SourcePortRange = "*",
                DestinationPortRange = "3389",
                SourceAddressPrefix = "VirtualNetwork",
                DestinationAddressPrefix = "VirtualNetwork",
                ResourceGroupName = resourceGroup.Name,
                NetworkSecurityGroupName = nsg.Name,
                Description = "Allows internal RDP Traffic"
            });


            var securityRules = new[]
            {
                new AzureNative.Network.Inputs.SecurityRuleArgs()
                {
                    Name = "NSG2-AllowVirtualNetworkRdpInbound",
                    Priority = 120,
                    Direction = SecurityRuleDirection.Inbound,
                    Access = SecurityRuleAccess.Allow,
                    Protocol = SecurityRuleProtocol.Tcp,
                    SourcePortRange = "*",
                    DestinationPortRange = "3389",
                    SourceAddressPrefix = "VirtualNetwork",
                    DestinationAddressPrefix = "VirtualNetwork",
                    Description = "Allows internal RDP Traffic"
                }
            };

            var nsg2 = new NetworkSecurityGroup("NSG2", new NetworkSecurityGroupArgs
            {
                NetworkSecurityGroupName = "NSG2",
                ResourceGroupName = resourceGroup.Name,
                Location = resourceGroup.Location,

                SecurityRules = securityRules
            });
        }
    }
}
```

**Comment by ie-zero** [comment]
The cause of this issue might be the ARM design as described by @sam-cogan at this https://github.com/pulumi/pulumi/issues/2247

**Comment by mikhailshilkov** [comment]
Sounds similar to https://github.com/pulumi/pulumi-azure-native/issues/611 but for a different resource. It's indeed a consequence of the ARM API design and our use of it.

While we are thinking of a structural solution, any chance you could define security rules inline in the security group?

**Comment by ie-zero** [comment]
I have to go back to the drawing board but it might be possible

**Comment by ie-zero** [comment]
@mikhailshilkov Thank you for looking into this so promptly.

**Comment by sy-bee** [comment]
Unfortunately there's an issue when an NSG is manually disassociated - it won't be added back on Pulumi run. I haven't tried running `refresh` as I'm using Automations API and do not explicitly call `refresh()` before running `up()`.

**Comment by sy-bee** [comment]
If anyone hits this issue, the workaround is to supply `tags` to `NetworkSecurityGroupArgs` which contain a value which changes on every run, e.g. `time.time()`. Here's sample:
```python3
security_group = azure_native.network.NetworkSecurityGroup(
    "nsg",
    location="eastus2",
    resource_group_name="rg",
)

# workaround for https://github.com/pulumi/pulumi-azure-native/issues/1087
nsg_tags = {"last_updated": str(int(time()))}
nsg_tags.update(self.tags)
subnet = azure_native.network.Subnet(
    "subnet-name",
    address_prefix="192.168.0.0/24",
    network_security_group=azure_native.network.NetworkSecurityGroupArgs(
        id=security_group.id,
        location="eastus2",
        tags=nsg_tags,
    ),
    resource_group_name="rg",
    virtual_network_name="vnet",
)
```

**Comment by tsbala** [comment]
We are facing this issue with `SecurityRules` being deleted and created in the `NetworkSecurityGroup` on successive runs of `pulumi up -r`.

Unfortunately, we cannot define the security rules inline when adding the `NetworkSecurityGroup`. We have also tried adding the `securityRules` to `IgnoreChanges` in the `NetworkSecurityGroup` but that doesn't make any difference.

Is there a workaround we can use to prevent the security rules from being deleted?

**Comment by danielrbradley** [comment]
This should now be resolved in [release v2.8.0](https://github.com/pulumi/pulumi-azure-native/releases/tag/v2.8.0) having merged:
- https://github.com/pulumi/pulumi-azure-native/pull/2755
